src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs

```diff
@@ -93,5 +97,22 @@ internal static int[] GetCurrencyNegativePatterns(CultureInfo cultureInfo)
 
             throw DateTimeFormatInfoData.GetCultureNotSupportedException(cultureInfo);
         }
+        
+        private static Version GetOSXKernelVersion()
+        {
+            if (s_isOSX)
+            {
+                ulong bytesLength = 256;
+                byte[] bytes = new byte[bytesLength];
+                sysctlbyname("kern.osrelease", bytes, ref bytesLength, null, 0);
+                string versionString = Encoding.UTF8.GetString(bytes);
+                return Version.Parse(versionString);
+            }
+            
+            return new Version(0, 0, 0);
+        }
+        
+        [DllImport("libc")]
+        private static extern int sysctlbyname(string ctlName, byte[] oldp, ref ulong oldpLen, byte[] newp, ulong newpLen);
```

Comment:
If I may interject...

I agree with @stephentoub: doing the same thing in the tests that you do in the code and then asserting that you get the same result is not a particularly valuable test. However, @tarekgh's point about the underlying data changing is also well taken. Choosing between "the lesser of two evils" seems to be the only choice in the current setup.

May I suggest that the reason for this difficulty is that these unit tests have a dependency on an external system over which the tests have no control: namely, the underlying OS. This results in a situation where the test does not control its input data, which is counter to the usual goals for unit tests (well-known inputs produce well-known outputs). The typical way to achieve this goal is to stub the external dependency to use well-known data; in this case, stub the underlying system calls to return representative data sets.

The downside of using such stubs is that the tests now only provide confidence that the code works given the well-known input data in the tests; there needs to be a corresponding confidence that the OS will actually supply data like what is in the tests in order to complete the chain. This can be accomplished using a form of contract test: a test that is specifically designed to validate that the external dependency returns the data (the "contract") that you expect in your unit tests. The key is that because these contract tests will only fail when the external dependency changes, and the external dependency in this case is the OS which will not change very often in CI, they don't need to run in CI. They only really need to be run any time the external dependency changes (which in the case of CoreFX would be when a new OS or OS version is added to the supported list).

This makes the tests much more valuable because it means that we can essentially test all of the expected scenarios from all OSs with normal CI _and_ local developer builds. Having CI for all possible OSs is not necessary, and a local developer doesn't have to wait for a full CI run to know if he has broken a scenario from an OS other than what is available to him.

Contract tests also address another point that @stephentoub brought up:

>> how do we write the regression test for the reported failure?

> We don't need to write a regression test if we fixed the issue.

If I understand correctly, inherent in @tarekgh's response is that you will know if you fixed the issue by running the tests _on the OS on which the issue was found_. By using contract tests, you can instead add a new contract to the unit tests that _represents_ that OS, then fix the issue, which will cause that test to pass, all on your local machine, regardless of its OS. The only thing you need the OS in question for is to establish the contract.

The downside to this approach is that running unit tests alone on an OS is not sufficient to determine whether that OS is fully supported by the framework. But that is simple to overcome: simply run the contract tests as well. In this way, the activities of everyday development are separated from the activity of supporting a new OS (or OS version), which IMHO is a good thing.

Stubbing the underlying OS is probably not a common approach; I don't know if there are any tests currently in CoreFX that take this approach. But for a framework with cross-platform ambitions like CoreFX, I think it is an approach worth considering.